hypershift-operator/controllers/hostedcluster/hostedcluster_webhook.go

```diff
@@ -69,6 +74,61 @@ func filterMutableHostedClusterSpecFields(spec *hyperv1.HostedClusterSpec) {
 	}
 }
 
+// validateStructDeepEqual walks through a struct and compares each entry.  If it comes accross a substruct it
+// recursively calls itself.  Returns a list of immutable field errors generated by any field being changed.
+func validateStructDeepEqual(x reflect.Value, y reflect.Value, path *field.Path, errs field.ErrorList) field.ErrorList {
+	for i := 0; i < x.NumField(); i++ {
+		v1 := x.Field(i)
+		v2 := y.Field(i)
+		json_id := x.Type().Field(i).Tag.Get("json")
+		sep := strings.Split(json_id, ",")
+		if len(sep) > 1 {
+			json_id = sep[0]
+		}
+
+		if v1.Kind() == reflect.Pointer {
+			// If this is a pointer to a struct, dereference before continuing.
+			if v1.Elem().Kind() == reflect.Struct {
+				v1 = v1.Elem()
+				v2 = v2.Elem()
+			}
+		}
+		if v1.Kind() == reflect.Struct {
+			errs = append(validateStructDeepEqual(v1, v2, path.Child(json_id), errs))
+		} else {
+			if v1.CanInterface() {
+				// Using DeepEqual() here because it takes care of all the type checking/comparison magic.
+				if !reflect.DeepEqual(v1.Interface(), v2.Interface()) {
```

Comment:
should use [equality.Semantic.DeepEqual()](https://github.com/openshift/hypershift/blob/122678ba9a81b147474bfad00479f263376b0671/control-plane-operator/controllers/awsprivatelink/awsprivatelink_controller.go#L292)